# 1. LinearRegression 모듈 이용 계산

In [18]:
import pandas as pd
import numpy as np

In [19]:
from sklearn.linear_model import LinearRegression

In [20]:
HW = pd.DataFrame({'SNS 시간': [10,80, 50, 40], '직전시험점수':[95,20,60,50], '시험점수':[90,10,85,50]})
HW

SNS 시간  직전시험점수  시험점수
0      10      95    90
1      80      20    10
2      50      60    85
3      40      50    50

In [21]:
X = HW.drop("시험점수", axis = 1, inplace = False)
y = HW["시험점수"]

In [22]:
model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [23]:
coef = pd.Series(data = model.coef_, index = X.columns)
print(coef)

SNS 시간    0.983834
직전시험점수    1.974596
dtype: float64


# 2. 직접 계산

In [24]:
import pandas as pd
import numpy as np
from numpy.linalg import inv
from numpy.linalg import pinv

In [25]:
HW = pd.DataFrame({'SNS 시간': [10,80, 50, 40], '직전시험점수':[95,20,60,50], '시험점수':[90,10,85,50]})
HW

SNS 시간  직전시험점수  시험점수
0      10      95    90
1      80      20    10
2      50      60    85
3      40      50    50

In [26]:
X = HW.drop("시험점수", axis = 1, inplace = False)
y = HW["시험점수"]

In [27]:
# Dataframe to numpy
X = pd.DataFrame(X).to_numpy()
y = pd.DataFrame(y).to_numpy()

In [28]:
# inv
Xpinv = inv(X.T @ X) @ X.T
print(Xpinv)

# pseufo inv
Xpinv2 = pinv(X)
print(Xpinv2)

[[-0.00522487  0.01014294  0.00300523  0.0022638 ]
 [ 0.00866008 -0.00364439  0.00240325  0.0021197 ]]
[[-0.00522487  0.01014294  0.00300523  0.0022638 ]
 [ 0.00866008 -0.00364439  0.00240325  0.0021197 ]]


In [29]:
x = Xpinv @ y

np.set_printoptions(precision=8, suppress=True)
print(x)

[[-0.00017432]
 [ 1.05322487]]


# 3. 비교

1. SNS 시간    0.983834
직전시험점수    1.974596
2. 직접 계산: [[-0.00017432]
 [ 1.05322487]]


# 4. 임의의 데이터 입력 후 예측값 구하기
| LinearRegression 활용

In [30]:
mydata = np.array([[60, 100]])

In [31]:
myscore = model.predict(mydata)
print('예상 점수: ', myscore)

예상 점수:  [159.8960739]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
